c/uca/sifter/unisifeggs.cpp

```diff
@@ -0,0 +1,440 @@
+        } else {
+            return it->second;
+        }
+    }
+
+    char32_t SimpleUppercaseMapping(char32_t const c) const {
+        auto const it = simple_uppercase_mapping_.find(c);
+        if (it == simple_uppercase_mapping_.end()) {
+            return c;
+        } else {
+            return it->second;
+        }
+    }
+
+    std::uint8_t CanonicalCombiningClass(char32_t const c) const {
+        auto const it = canonical_combining_class_.find(c);
+        if (it == canonical_combining_class_.end()) {
+            return 0;
+        } else {
+            return it->second;
+        }
+    }
+
+    std::optional<std::uint64_t> NaturalNumericValue(char32_t const c) const {
+        auto const it = natural_numeric_value_.find(c);
+        if (it == natural_numeric_value_.end()) {
+            return std::nullopt;
+        } else {
+            return it->second;
+        }
+    }
+
+    CodePointSet const& CodePointsWithNumericValue() const {
+        return code_points_with_numeric_value_;
+    }
+
+   private:
+    void ReadMultiPropertyFile(std::filesystem::path const& file) {
+        std::ifstream prop_list(file);
+        CHECK(prop_list.good());
+        std::cout << "Reading " << file << "...\n";
+        for (std::string line; std::getline(prop_list, line);) {
+            // See https://www.unicode.org/reports/tr44/#UnicodeData.txt.
+            auto const record = strip_comment(line);
+            if (record.empty()) {
+                continue;
+            }
+            switch (std::count(record.begin(), record.end(), ';')) {
+                case 1: {
+                    auto const [cp, property_name] = fields<2>(record);
+                    CodePointRange const range = parse_hex_codepoint_range(cp);
+                    binary_properties_[std::string(property_name)].Include(
+                        range);
+                    break;
+                }
+                case 2:  // Enumerated property.
+                    break;
+                default:
+                    std::terminate();
+            }
+        }
+    }
+
+    // Keys are C, L, M, N, P, S, Z.
+    std::map<char, CodePointSet> coarse_general_category_;
+    // Keyed on short alias.
+    std::map<std::string, CodePointSet, std::less<>> general_category_;
+    std::map<std::string, CodePointSet, std::less<>> binary_properties_;
+    // Default <code point> omitted.
+    std::map<char32_t, char32_t> simple_uppercase_mapping_;
+    // Default <code point> omitted.
+    std::map<char32_t, char32_t> simple_lowercase_mapping_;
+    // Default 0 omitted.
+    std::map<char32_t, std::uint8_t> canonical_combining_class_;
+    std::map<char32_t, std::uint64_t> natural_numeric_value_;
+    // Characters with a non-NaN Numeric_Value.
+    CodePointSet code_points_with_numeric_value_;
+};
+
+typedef unsigned short UShort16;
+
+typedef unsigned int UInt32;
+
+UCD* const ucd = new UCD();
+
+static bool isSifterNonSpacingCombining(char32_t c) {
```

Comment:
Rather than "static", I would put everything in this file after the includes and before the `extern "C"` functions into an anonymous namespace:

```
namespace {

// no additional indent

}  // namespace
```